### PR TITLE
KMS UPDATE: non hardware based operations

### PR DESCRIPTION
**Pull request by ehanoc**
# What

- Issues when no hardware support is available for cryptographic / kms operations
- Isolation strategies
- Run-time environments security risks
- Mitigating risks with app / kms designs

### CODE REVIEW

File: docs/architecture/key-management-services.md [hunk 1]
```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
...
+The examples above are conceptual designs and not **prescriptive**. We need to consider the specifics of the OS we are running on. For example,
+
+- Android
+    - Supports isolated services
+    - Lifecycle control
+    - Natural sandboxing between apps
+    - TEE + HSM available for at-rest / transit encryption
+- iOS
+    - Isolated services seem to be only possible through Scheduled Tasks (https://developer.apple.com/documentation/backgroundtasks) 
+    - Besides the traditional NIST crypto; Apple has native support for Curve25519,EdDSA, ECDH, xchacha20 symmetric encryption scheme
```
Comment (c00kiemon5ter) [inline docs/architecture/key-management-services.md:252]:
```suggestion
    - Besides the traditional NIST crypto; Apple has native support for Curve25519, EdDSA, ECDH, xchacha20 symmetric encryption scheme
```

File: docs/architecture/key-management-services.md [hunk 2]
```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
...
+        title "Isolation KMS, App, TEE and hardware"
+            Boundary(b0, "Android", "OS") {
+            
+                Boundary(b1, "Application", "Main process") {
+                    System(ow, "OWF Engine", "Wallet")
+                }
+                
+                Boundary(b2, "KMS", "ISOLATED SERVICE") {
+                    Component(ipcClientHandler, "KMS Service", "Service", "Isolated kms")
+                    ComponentDb(kmsDb, "keyStorage", "Database","Encrypted Keystore")
```
Comment (c00kiemon5ter) [inline docs/architecture/key-management-services.md:222]:
```suggestion
                    ComponentDb(kmsDb, "keyStorage", "Database", "Encrypted Keystore")
```

File: docs/architecture/key-management-services.md [hunk 3]
```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
...
+                }
+                
+                Component(tee, "TEE", "TEE","Trusted Execution Environment")
+                Component(hsm, "HSM", "HSM","Hardware Security Module")
```
Comment (c00kiemon5ter) [inline docs/architecture/key-management-services.md:226]:
```suggestion
                Component(tee, "TEE", "TEE", "Trusted Execution Environment")
                Component(hsm, "HSM", "HSM", "Hardware Security Module")
```

**Review by c00kiemon5ter — CHANGES_REQUESTED** [review]
Minor formatting/whitespace changes.

In general, I like this.

File: docs/architecture/key-management-services.md [hunk 4]
```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
...
+- Use device bound hardware for encryption / decryption
+- At rest encryption with native support through TEE & phone’s native HSM 
+- At run-time decrypt only necessary keys
+- The application is blind to any private key material and should only request operations from the KMS. Operations can be (not excluded to):
+    - Key-gen
+    - Get-publicKey
+    - Sign
+    - Key-derive
+    - Key-import 
+    - etc
```
Comment (c00kiemon5ter) [inline docs/architecture/key-management-services.md:200]: I am sure we are not the first to propose such a separation.
Can we have examples of other systems working this way?
Can we learn from their experience?
Can we study and reuse their APIs?

File: docs/architecture/key-management-services.md [hunk 5]
```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
 
 ```
 
+### Non-Hardware Operations
+
+Given that a wallet maybe be deployed or installed in a OS or environment that does not have hardware available that can provide all the cryptographic operations needed, we need to consider a non-hardware mode of operation.
```
Comment (tkuhrt) [inline docs/architecture/key-management-services.md:117]:
```suggestion
Given that a wallet may be deployed or installed in a OS or environment that does not have hardware available that can provide all the cryptographic operations needed, we need to consider a non-hardware mode of operation.
```

File: docs/architecture/key-management-services.md [hunk 6]
```diff
@@ -108,4 +112,143 @@ There can be some variations in the support of these crypto primitives, hashing
...
+
+        UpdateRelStyle(app, kms, $textColor="green", $lineColor="blue", $offsetX="10")
+        UpdateLayoutConfig($c4ShapeInRow="1", $c4BoundaryInRow="0")
+```
+
+##### Problem with Shared Run-time space
+- Shared resources
+- Permissions to the app open attack surface
+- No compartmentalization of a breach or zero-day exploit
+Non compliant with some security industry standards and regulations where separation of duties is a must
```
Comment (tkuhrt) [inline docs/architecture/key-management-services.md:146]:
```suggestion
- Non compliant with some security industry standards and regulations where separation of duties is a must
```

**Pull request merged**